### DISCUSSION TITLE
VSCode and jsonargparse error

### DISCUSSION THREAD

**Discussion by nicoloesch** [Lightning-AI/pytorch-lightning#18025] [Lightning Trainer API: Trainer, LightningModule, LightningDataModule] [answered]
Hi everyone, 

I recently switchted to VSCode from PyCharm and I am facing currently an issue. If I try to debug my `main.py` file, which uses the `LightningCLI` (`DiffLightningCLi` inherits from it and just overwrites `add_arguments_to_parser`):

```python
import sys
import os

from adiff.utils import train_util
import adiff.model.lightning.base_model


def cli_main():
    # Start Training/Testing based on the args

      data_dir = f'/home/{os.getlogin()}/Data/datasets'
      logging_dir = f'/home/{os.getlogin()}/Data/lightning_logs'
      default_cfg = ["config/debug_cfg.yaml"]
      cmdline_args = [
          "fit", 
              "--model", "DiffGen",
              '--task', 'GEN',
              '--data_dir', f'{data_dir}', 
              '--logging_dir', f'{logging_dir}',
              '--data.dimensions', '2',
              '--num_workers', '0',
              '--data.batch_size=5',
              '--data.dataset=cifar-10',
              '--data.in_channels=3',
              '--model.out_channels=3',
              '--trainer.logger=False',
              '--data.num_train_subj=10',
              '--data.num_val_subj=10',
              '--model.diffusion_mean_type=EPSILON',
              '--model.diffusion_loss_type=SIMPLE',
              '--model.diffusion_var_type=FIXED_LARGE',
              '--model.ema_decay=0.9999',
              '--trainer.max_steps=700000',
              ]
    subcommands = ('fit', 'test', 'validate', 'predict')
    parser_kwargs = {subcommand: {'default_config_files': default_cfg} for subcommand in subcommands}

    cli = train_util.DiffLightningCLI(
            model_class=adiff.model.lightning.base_model.BaseModel,
            subclass_mode_model=True,
            run=run,
            seed_everything_default=123,
            parser_kwargs=parser_kwargs,
            save_config_kwargs={"overwrite": True,
                                "multifile": True},
            args=cmdline_args)

    # Shutdown
    sys.exit()


if __name__ == "__main__":
    import warnings
    from lightning_fabric.utilities.warnings import PossibleUserWarning

    with warnings.catch_warnings():
        warnings.filterwarnings(
            action='ignore',
            category=PossibleUserWarning,
            message='The dataloader,')
        cli_main()

```
**I am aware that passing in the args like this is not recommended and is also only used in debugging and not in the final `main.py`**. Running/debugging the script using the `launch.json` configuration of VSCode 
```json
{
    "version": "0.2.0",
    "configurations": [
        {
            "name": "main.py",
            "type": "python",
            "request": "launch",
            "program": "${workspaceFolder}/ADiff/adiff/main.py",
            "console": "integratedTerminal",
            "justMyCode": false,
            "redirectOutput": false,
        }
    ]
}
```
I get the following error message:

```
main.py [options] fit: error: Parser key "data":
  Not a valid subclass of LightningDataModule. Got value: NestedArg(key='dimensions', val='2')
  Subclass types expect one of:
  - a class path (str)
  - a dict with class_path entry
  - a dict without class_path but with init_args entry (class path given previously)
 ```

The command under the hood in VSCode uses `cd /home/u24417317/Code ; /usr/bin/env /home/u24417317/.conda/envs/phd39/bin/python /home/u24417317/.vscode-server/extensions/ms-python.python-2023.12.0/pythonFiles/lib/python/debugpy/adapter/../../debugpy/launcher 60221 -- /home/u24417317/Code/ADiff/adiff/main.py` so the correct file is executed.

However, if I run the exact same file using python directly without VSCode, the code runs as intended (as it was running on PyCharm).
```
(phd39) u24417317@LabServer:~/Code/ADiff/adiff$ python main.py
Global seed set to 123
GPU available: True (cuda), used: True
TPU available: False, using: 0 TPU cores
IPU available: False, using: 0 IPUs
HPU available: False, using: 0 HPUs
```
I am debugging remotely using ssh (I am running the same file with and without VSCode), made sure that the exact same conda environment is utilised with the following packages:

jsonargparse: 4.22.1
lightning-utilities: 0.8.0
pytorch-lightning: 2.0.4

It seems to me like a VSCode issue, hence why I am aware that this might not be the best place to ask that question but maybe someone has encountered a similar problem or could share their working `launch.json` configuration.

I appreciate any input!
Nico

**Comment by mauvilsa** [comment]
I think this is not related to VSCode. I can reproduce the same when running from command line with:

```python
from lightning.pytorch.cli import LightningCLI
from lightning.pytorch.demos.boring_classes import DemoModel

def cli_main():
    cmdline_args = [
        "fit", 
            "--model", "DemoModel",
            '--data.dimensions', '2',
    ]
    cli = LightningCLI(
        model_class=DemoModel,
        subclass_mode_model=True,
        run=True,
        args=cmdline_args)

if __name__ == "__main__":
    cli_main()
```

The problem is "a dict without class_path but with init_args entry (class path given previously)". That is, `--data.dimensions=2` is given, but it is not known which data class to check if it accepts a `dimensions` parameter. The error goes away if before that argument, there is `--data=SomeDataClass`.

The error message could be better. To request to improve that, please do in https://github.com/omni-us/jsonargparse/issues.

**Reply by nicoloesch** [reply]
@mauvilsa thanks for the information! I forgot to mention that I provide the respective class with `init_args` in a `default_cfg.yaml`. As such, the class should be fully instantiated with the default arguments and then with the manually placed args (reflecting command line args) overwritten for debugging.

The main problem I have is the following: if I try to debug `main.py` through VSCode without any other commands, I get the error message. If I execute `main.py` just using python with the same conda environment, the code runs as expected (and as it has been doing using PyCharm). As a result, there **MUST** be a difference running `main.py` through VSCode and without it.

```.yaml
# This is the configuration file for all models inheriting from BaseModel
# It is specified here as class path despite not requiring most of the args
# as in the instantiation of the CLI, I specified model_class as the BaseModel
# and allow the subclass_mode_model
model:
  class_path: adiff.model.lightning.base_model.BaseModel
  init_args:
    activation: 'ReLU'
    act_args:
      - false
    attention: true
    attention_heads: 1
    attention_res:
      - 16
    channel_factor:
      - 1
      - 2
      - 2
      - 2
    classifier_ckpt: null
    concat_sc: true
    conv_zero_init: false
    diffusion_gradient_scale: 1.0
    diffusion_log_var: true
    diffusion_loss_type: SIMPLE
    diffusion_mean_type: EPSILON
    diffusion_noising: 'linear'
    diffusion_steps: 1000
    diffusion_var_type: FIXED_LARGE
    diffusion_verbose: true
    dropout: 10
    hidden_channels: 128
    kernel_size: 3
    learning_rate: 1e-4
    norm: 'group'
    num_fid_samples: 50000
    num_res_blocks: 2
    out_channels: 4
    padding: true
    padding_mode: 'zeros'
    preactivation: true
    spatial_factor: 2
    upd_conv: true
    ema_decay: 0.0
data:
  class_path: adiff.input.data_loader.DataModuleBraTS
  init_args:
    augment: false
    batch_size: 2
    brats_2023_subtask: null
    conditional_sampling: false
    #dataset: 'cifar-10'
    dataset: 'BraTS_2020'
    dimensions: 3
    fg_prob: 3
    in_channels: 4
    multiclass_pred: true
    num_train_subj: 2  # null enables the utilisation of 60%, otherwise explicit
    num_val_subj: 2  # null enables the utilisation of 20%, otherwise explicit
    num_test_subj: 2  # null enables the utilisation of 20%, otherwise explicit
    patches_per_subj: 1
    patch_overlap: 8
    patch_size: 32
    resample: false
    to_ras: false
trainer:
  accelerator: auto
  check_val_every_n_epoch: 1
  detect_anomaly: true
  devices: auto
  deterministic: true
  benchmark: false
  enable_checkpointing: true
  enable_progress_bar: true
  enable_model_summary: true
  fast_dev_run: false
  gradient_clip_val: null
  gradient_clip_algorithm: 'norm'
  inference_mode: true
  limit_train_batches: null
  limit_val_batches: null
  limit_test_batches: null
  limit_predict_batches: null
  log_every_n_steps: 1
  logger:
    class_path: adiff.utils.callbacks.Logger
  max_epochs: 5
  max_steps: -1
  min_epochs: 1
  num_nodes: 1
  num_sanity_val_steps: 0
  precision: 32-true
  strategy: auto
  use_distributed_sampler: false
progress_bar:
  refresh_rate: 1
  process_position: 0
#optimizer:
#  class_path: torch.optim.Adam
#  init_args:
#    lr: 1e-5
#    weight_decay: 0.0
#lr_scheduler:
#  class_path: adiff.utils.callbacks.WarmUpScheduler
#  init_args:
#    warmup: 20
```

**Reply by nicoloesch** [reply, marked as answer]
@mauvilsa I figured it out. It turned out that my default configuration file was not found due to an error in the `launch.json` file of VSCode. I provided the configuration as a relative path to the `main.py` file. However, the `cwd` was set to the `${workspaceFolder}`, which is two folders above it (as I want to have all my different projects in the same workspace). This lead to not finding the respective default configuration file in `_get_default_config_files` of `jsonargparse/core.py`. 
Your suggestion of having a more distinct error message could be helpful in the future but is included in the current error messages if one has more experience with the `jsonargparse` module.

Thank you for your help! 
Nico

-------------
Solution: Setting
`"cwd": "${workspaceFolder}/<path_to_main_file>"` in `launch.json`
